File: cmd/inbox.go
package cmd

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/dramxx/later/config"
	"github.com/dramxx/later/gist"
)

func buildContent(nonEntryLines, entryLines []string) string {
	var lines []string
	for _, line := range nonEntryLines {
		if line != "" {
			lines = append(lines, line)
		}
	}
	for _, line := range entryLines {
		lines = append(lines, line)
	}
	if len(lines) == 0 {
		return ""
	}
	return strings.Join(lines, "\n") + "\n"
}

func Inbox() {
	cfg, err := config.Load()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	content, err := gist.GetInbox(cfg)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	lines := strings.Split(content, "\n")
	if len(lines) == 0 || (len(lines) == 1 && lines[0] == "") {
		lines = nil
	}
	var entryLines []string
	var nonEntryLines []string

	for _, line := range lines {
		if strings.HasPrefix(line, "[") {
			entryLines = append(entryLines, line)
		} else {
			nonEntryLines = append(nonEntryLines, line)
		}
	}

	if len(os.Args) >= 3 && os.Args[2] == "--clear" {
		newContent := buildContent(nonEntryLines, nil)

		err = gist.UpdateInbox(cfg, newContent)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
			os.Exit(1)
		}
		fmt.Println("✓ cleared")
		return
	}

	if len(os.Args) >= 3 && os.Args[2] == "--pop" {
		if len(os.Args) < 4 {
			fmt.Println("Usage: later inbox --pop <n> [n...]")
			os.Exit(1)
		}

		indices := []int{}
		seen := make(map[int]bool)
		for _, arg := range os.Args[3:] {
			n, err := strconv.Atoi(arg)
			if err != nil || n < 1 || n > len(entryLines) {
				fmt.Fprintf(os.Stderr, "Error: invalid index: %s\n", arg)
				os.Exit(1)
			}
			if seen[n] {
				fmt.Fprintf(os.Stderr, "Error: duplicate index: %d\n", n)
				os.Exit(1)
			}
			seen[n] = true
			indices = append(indices, n)
		}

		removed := 0
		for i := len(entryLines) - 1; i >= 0; i-- {
			for _, idx := range indices {
				if i == idx-1 {
					entryLines = append(entryLines[:i], entryLines[i+1:]...)
					removed++
					break
				}
			}
		}

		newContent := buildContent(nonEntryLines, entryLines)

		err = gist.UpdateInbox(cfg, newContent)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
			os.Exit(1)
		}

		if removed == 1 {
			fmt.Println("✓ removed 1 entry")
		} else {
			fmt.Printf("✓ removed %d entries\n", removed)
		}
		return
	}

	if len(entryLines) == 0 {
		fmt.Println("inbox is empty")
		return
	}

	for i, line := range entryLines {
		fmt.Printf("%d  %s\n", i+1, line)
	}
}

File: src/commands/inbox.rs
use crate::config;
use crate::gist;
use std::env;

pub fn inbox() {
    let args: Vec<String> = env::args().collect();

    let cfg = match config::load() {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    };

    let content = match gist::get_inbox(&cfg) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
    };

    let lines: Vec<&str> = content.lines().collect();
    let (non_entry_lines, entry_lines): (Vec<&str>, Vec<&str>) =
        lines.iter().partition(|l| !l.starts_with('['));

    if args.len() >= 3 && args[2] == "--clear" {
        let new_content: String = non_entry_lines
            .iter()
            .filter(|l| !l.is_empty())
            .map(|l| *l)
            .collect::<Vec<_>>()
            .join("\n");
        let final_content = if new_content.is_empty() {
            String::new()
        } else {
            new_content + "\n"
        };

        if let Err(e) = gist::update_inbox(&cfg, &final_content) {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }
        println!("✓ cleared");
        return;
    }

    if args.len() >= 3 && args[2] == "--pop" {
        if args.len() < 4 {
            eprintln!("Usage: later inbox --pop <n> [n...]");
            std::process::exit(1);
        }

        let indices: Result<Vec<usize>, _> = args[3..].iter().map(|s| s.parse::<usize>()).collect();
        let indices = indices.unwrap_or_else(|_| {
            eprintln!("Error: invalid index");
            std::process::exit(1);
        });

        for &idx in &indices {
            if idx == 0 || idx > entry_lines.len() {
                eprintln!("Error: invalid index: {}", idx);
                std::process::exit(1);
            }
        }

        let new_entries: Vec<&str> = entry_lines
            .iter()
            .enumerate()
            .filter(|(i, _)| !indices.contains(&(i + 1)))
            .map(|(_, &l)| l)
            .collect();

        let new_content: String = non_entry_lines
            .iter()
            .filter(|l| !l.is_empty())
            .chain(new_entries.iter())
            .map(|&l| l)
            .collect::<Vec<_>>()
            .join("\n");
        let final_content = if new_content.is_empty() {
            String::new()
        } else {
            new_content + "\n"
        };

        if let Err(e) = gist::update_inbox(&cfg, &final_content) {
            eprintln!("Error: {}", e);
            std::process::exit(1);
        }

        let removed = indices.len();
        if removed == 1 {
            println!("✓ removed 1 entry");
        } else {
            println!("✓ removed {} entries", removed);
        }
        return;
    }

    if entry_lines.is_empty() {
        println!("inbox is empty");
        return;
    }

    for (i, line) in entry_lines.iter().enumerate() {
        println!("{}  {}", i + 1, line);
    }
}
